File: analysis/a11y/schema.py
from dataclasses import dataclass
from typing import Any, Literal, Optional

Tool = Literal['axe-core', 'ibm-equal-access']


@dataclass
class Violation:
    tool: Tool
    rule_id: str
    wcag: list[str]
    description: str
    count: int
    targets: list[str]
    impact: Optional[str] = None
    help_url: Optional[str] = None


@dataclass
class Metadata:
    timestamp: str
    url: str
    browser: str
    browser_version: str
    viewport: Optional[dict[str, int]]
    device: str
    tool: Tool
    tool_version: str
    wcag_tags: Optional[list[str]] = None


@dataclass
class Result:
    metadata: Metadata
    violations: list[Violation]
    raw: Any

File: collection/schema.ts
export type Tool = 'axe-core' | 'ibm-equal-access'

export interface Violation {
  tool: Tool
  ruleId: string
  wcag: string[]
  description: string
  count: number
  targets: string[]
  impact?: string
  helpUrl?: string
}

export interface Metadata {
  timestamp: string
  url: string
  browser: string
  browserVersion: string
  viewport: { width: number; height: number } | null
  device: string
  tool: Tool
  toolVersion: string
  wcagTags?: readonly string[]
}

export interface Result<Raw = unknown> {
  metadata: Metadata
  violations: Violation[]
  raw: Raw
}

export interface MissingAsset {
  status: number
  resourceType: string
  url: string
}

export interface Fidelity {
  stylesheets: number
  missingAssets: MissingAsset[]
}
